File: sample_source_code/CPP/tests/ShelterTest.cpp
#include <gtest/gtest.h>
#include "Shelter.h"
#include "Dog.h"
#include "Cat.h"

class ShelterTest : public ::testing::Test {
protected:
    Shelter shelter{"Happy Paws", 5};
    std::shared_ptr<Dog> rex      = std::make_shared<Dog>("Rex",      3, "Labrador", 30.0);
    std::shared_ptr<Cat> whiskers = std::make_shared<Cat>("Whiskers", 2, true,        4.5);
};

// ── getName ───────────────────────────────────────────────────────────────────

TEST_F(ShelterTest, GetName_ReturnsConstructorValue) {
    EXPECT_EQ("Happy Paws", shelter.getName());
}

// ── intake ────────────────────────────────────────────────────────────────────

TEST_F(ShelterTest, Intake_UnderCapacity_ReturnsTrue) {
    EXPECT_TRUE(shelter.intake(rex));
}

TEST_F(ShelterTest, Intake_AddsAnimalToList) {
    shelter.intake(rex);
    EXPECT_EQ(1u, shelter.getAnimals().size());
}

TEST_F(ShelterTest, Intake_AtCapacity_ReturnsFalse) {
    Shelter tiny{"Tiny", 1};
    tiny.intake(rex);
    EXPECT_FALSE(tiny.intake(whiskers));
}

TEST_F(ShelterTest, Intake_AtCapacity_DoesNotAddAnimal) {
    Shelter tiny{"Tiny", 1};
    tiny.intake(rex);
    tiny.intake(whiskers);
    EXPECT_EQ(1u, tiny.getAnimals().size());
}

TEST_F(ShelterTest, Intake_MultipleAnimals_AllAdded) {
    shelter.intake(rex);
    shelter.intake(whiskers);
    EXPECT_EQ(2u, shelter.getAnimals().size());
}

// ── adopt ─────────────────────────────────────────────────────────────────────

TEST_F(ShelterTest, Adopt_ExistingAnimal_ReturnsAnimal) {
    shelter.intake(rex);
    auto adopted = shelter.adopt("Rex");
    ASSERT_NE(nullptr, adopted);
    EXPECT_EQ("Rex", adopted->getName());
}

TEST_F(ShelterTest, Adopt_ExistingAnimal_RemovesFromShelter) {
    shelter.intake(rex);
    shelter.adopt("Rex");
    EXPECT_TRUE(shelter.getAnimals().empty());
}

TEST_F(ShelterTest, Adopt_UnknownName_ReturnsNullptr) {
    shelter.intake(rex);
    EXPECT_EQ(nullptr, shelter.adopt("Unknown"));
}

TEST_F(ShelterTest, Adopt_UnknownName_DoesNotRemoveOtherAnimals) {
    shelter.intake(rex);
    shelter.adopt("Unknown");
    EXPECT_EQ(1u, shelter.getAnimals().size());
}

// ── availableCapacity ─────────────────────────────────────────────────────────

TEST_F(ShelterTest, AvailableCapacity_EmptyShelter_EqualsCapacity) {
    EXPECT_EQ(5, shelter.availableCapacity());
}

TEST_F(ShelterTest, AvailableCapacity_AfterOneIntake_Decreases) {
    shelter.intake(rex);
    EXPECT_EQ(4, shelter.availableCapacity());
}

TEST_F(ShelterTest, AvailableCapacity_AfterAdopt_Increases) {
    shelter.intake(rex);
    shelter.adopt("Rex");
    EXPECT_EQ(5, shelter.availableCapacity());
}

// ── averageAge ────────────────────────────────────────────────────────────────

TEST_F(ShelterTest, AverageAge_EmptyShelter_ReturnsZero) {
    EXPECT_DOUBLE_EQ(0.0, shelter.averageAge());
}

TEST_F(ShelterTest, AverageAge_SingleAnimal_ReturnsItsAge) {
    shelter.intake(rex);   // age 3
    EXPECT_DOUBLE_EQ(3.0, shelter.averageAge());
}

TEST_F(ShelterTest, AverageAge_MultipleAnimals_IsCorrect) {
    shelter.intake(rex);       // age 3
    shelter.intake(whiskers);  // age 2
    EXPECT_NEAR(2.5, shelter.averageAge(), 1e-9);
}

File: sample_source_code/javascript/tests/shelter.test.js
'use strict';

const Shelter = require('../src/Shelter');
const Dog     = require('../src/Dog');
const Cat     = require('../src/Cat');

describe('Shelter.intake', () => {
  test('intake returns true and adds animal', () => {
    const shelter = new Shelter('Test', 5);
    expect(shelter.intake(new Dog('Rex', 3, 'Labrador', 30))).toBe(true);
    expect(shelter.getAnimals()).toHaveLength(1);
  });
  test('intake returns false when at capacity', () => {
    const shelter = new Shelter('Small', 1);
    shelter.intake(new Dog('Rex', 3, 'Labrador', 30));
    expect(shelter.intake(new Dog('Max', 5, 'Labrador', 30))).toBe(false);
    expect(shelter.getAnimals()).toHaveLength(1);
  });
  test('multiple intakes up to capacity all succeed', () => {
    const shelter = new Shelter('Medium', 3);
    expect(shelter.intake(new Dog('A', 1, 'Labrador', 30))).toBe(true);
    expect(shelter.intake(new Dog('B', 2, 'Labrador', 30))).toBe(true);
    expect(shelter.intake(new Dog('C', 3, 'Labrador', 30))).toBe(true);
    expect(shelter.getAnimals()).toHaveLength(3);
  });
});

describe('Shelter.adopt', () => {
  let shelter;

  beforeEach(() => {
    shelter = new Shelter('Test', 10);
    shelter.intake(new Dog('Rex', 3, 'Labrador', 30));
    shelter.intake(new Cat('Luna', 2, true, 4.0));
  });

  test('adopt returns the correct animal', () => {
    const adopted = shelter.adopt('Rex');
    expect(adopted).not.toBeNull();
    expect(adopted.getName()).toBe('Rex');
  });
  test('adopted animal is removed from shelter', () => {
    shelter.adopt('Rex');
    expect(shelter.getAnimals()).toHaveLength(1);
    expect(shelter.getAnimals()[0].getName()).toBe('Luna');
  });
  test('adopt is case-insensitive', () => {
    const adopted = shelter.adopt('rex');
    expect(adopted.getName()).toBe('Rex');
  });
  test('adopt returns null for unknown name', () => {
    expect(shelter.adopt('Unknown')).toBeNull();
  });
  test('shelter unchanged after failed adopt', () => {
    shelter.adopt('Unknown');
    expect(shelter.getAnimals()).toHaveLength(2);
  });
});

describe('Shelter.getAnimals', () => {
  test('returns a defensive copy — modifying it does not affect shelter', () => {
    const shelter = new Shelter('Test', 10);
    shelter.intake(new Dog('Rex', 3, 'Labrador', 30));
    const copy = shelter.getAnimals();
    copy.pop();
    expect(shelter.getAnimals()).toHaveLength(1);
  });
  test('returns empty array for new shelter', () => {
    expect(new Shelter('Empty', 10).getAnimals()).toHaveLength(0);
  });
});

describe('Shelter.availableCapacity', () => {
  test('full capacity available when empty', () => {
    expect(new Shelter('Test', 10).availableCapacity()).toBe(10);
  });
  test('decreases by 1 per intake', () => {
    const shelter = new Shelter('Test', 10);
    shelter.intake(new Dog('Rex', 3, 'Labrador', 30));
    expect(shelter.availableCapacity()).toBe(9);
  });
  test('increases by 1 after successful adopt', () => {
    const shelter = new Shelter('Test', 10);
    shelter.intake(new Dog('Rex', 3, 'Labrador', 30));
    shelter.adopt('Rex');
    expect(shelter.availableCapacity()).toBe(10);
  });
});

describe('Shelter.averageAge', () => {
  test('returns 0 for empty shelter', () => {
    expect(new Shelter('Empty', 10).averageAge()).toBe(0);
  });
  test('returns age of single animal', () => {
    const shelter = new Shelter('Test', 10);
    shelter.intake(new Dog('Rex', 4, 'Labrador', 30));
    expect(shelter.averageAge()).toBe(4);
  });
  test('returns correct mean age for multiple animals', () => {
    const shelter = new Shelter('Test', 10);
    shelter.intake(new Dog('Rex',  4, 'Labrador', 30));
    shelter.intake(new Cat('Luna', 2, true, 4.0));
    expect(shelter.averageAge()).toBe(3);
  });
});
